### PR TITLE
replace transmute by raw pointer cast

### PR DESCRIPTION
**Pull request by RalfJung**
Now that https://github.com/rust-lang/rust/issues/113257 is fixed we can finally do this. :)

### CODE REVIEW

**Comment by thomcc** [comment]
LGTM

@bors r+

**Comment by thomcc** [comment]
r=me with green ci.

File: library/std/src/thread/mod.rs [hunk 1]
```diff
@@ -545,6 +545,13 @@ impl Builder {
...
 
+        let main = Box::new(main);
+        #[cfg(bootstrap)]
+        let main =
+            unsafe { mem::transmute::<Box<dyn FnOnce() + 'a>, Box<dyn FnOnce() + 'static>>(main) };
+        #[cfg(not(bootstrap))]
+        let main = unsafe { Box::from_raw(Box::into_raw(main) as *mut (dyn FnOnce() + 'static)) };
+
         Ok(JoinInner {
             // SAFETY:
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/std/src/thread/mod.rs:556]: Not quite, that comment is specific to what `imp::Thread::new` does.. but we should probably forward-reference it from here: https://github.com/rust-lang/rust/pull/117281

**Pull request merged**